batchflow/models/tf/layers/conv_block.py

```diff
@@ -159,33 +143,296 @@ def conv_block(inputs, layout='', filters=0, kernel_size=3, name=None,
+
+    def __init__(self, layout='', filters=0, kernel_size=3, name=None,
+                 strides=1, padding='same', data_format='channels_last', dilation_rate=1, depth_multiplier=1,
+                 activation=tf.nn.relu, pool_size=2, pool_strides=2, dropout_rate=0., **kwargs):
+        self.layout = layout
+        self.filters, self.kernel_size, self.strides = filters, kernel_size, strides
+        self.padding, self.data_format, self.name = padding, data_format, name
+        self.dilation_rate, self.depth_multiplier = dilation_rate, depth_multiplier
+        self.pool_size, self.pool_strides = pool_size, pool_strides
+        self.activation = activation
+        self.dropout_rate = dropout_rate
+        self.kwargs = kwargs
+
+
+    def add_letter(self, letter, func, name=None):
+        """ Add custom letter to layout parsing procedure.
+
+        Parameters
+        ----------
+        letter : str
+            Letter to add.
+        func : class
+            Tensor-processing layer. Must have layer-like signature (both init and call overloaded).
+        name : str
+            Name of parameter dictionary. Defaults to `letter`.
+
+        Examples
+        --------
+        Add custom `Q` letter::
+
+            block = ConvBlock('cnap Q', filters=32, custom_params={'key': 'value'})
+            block.add_letter('Q', my_func, 'custom_params')
+            x = block(x)
+        """
+        name = name or letter
+        self.C_LAYERS.update({letter: name})
+        self.FUNC_LAYERS.update({name: func})
+        self.C_GROUPS.update({letter: letter})
+
+
+    def __call__(self, inputs, training=None):
+        if training is None:
+            training = self.kwargs.get('is_training')
+        if training is None:
+            training = self.kwargs.get('training')
+
+        layout = self.layout or ''
+        layout = layout.replace(' ', '')
+
+        if len(layout) == 0:
+            logger.warning('ConvBlock: layout is empty, so there is nothing to do, just returning inputs.')
+            return inputs
+
+        context = None
+        if self.name is not None:
+            context = tf.variable_scope(self.name, reuse=self.kwargs.get('reuse'))
+            context.__enter__()
+
+        layout_dict = {}
+        for layer in layout:
+            if self.C_GROUPS[layer] not in layout_dict:
+                layout_dict[self.C_GROUPS[layer]] = [-1, 0]
+            layout_dict[self.C_GROUPS[layer]][1] += 1
+
+
+        residuals = []
+        tensor = inputs
+        for i, layer in enumerate(layout):
+
+            layout_dict[self.C_GROUPS[layer]][0] += 1
+            layer_name = self.C_LAYERS[layer]
+            layer_fn = self.FUNC_LAYERS[layer_name]
+
+            args = {}
+            call_args = {}
+
+            if layer == 'a':
+                args = dict(activation=self.activation)
+                layer_fn = unpack_args(args, *layout_dict[self.C_GROUPS[layer]])['activation']
```

Comment:
refactoring: 
1) add `layout_dict[self.C_GROUPS[layer]]` as a variable, because it is used multiple times
2) rename `layer_fn` to `activation_fn` or something like this, because now it shadows upper `layer_fn`

Reply:
Done